File: internal/tests/nodes_js_to_test/nodeMidToTest.ts
import { FinishInterface, FlowInterface, NodeInterface } from "./type"
import {writeFile} from 'fs/promises'


const baseURL = "http://localhost:8080/api/flow-state"
//process.env.FLOW_ID = "b620fec7-19bc-4c36-957d-248e666410c3"
//process.env.NODE_ID = "f50c3f51-8871-4e77-8132-bd5c8245bbfb"


function getNode(nodes: NodeInterface[], nodeId: string) {
    return nodes.find(node => node.id === nodeId)   
}

async function getFlowState() {
    const flowId = process.env.FLOW_ID
    const url = baseURL + `/get-flow-state/${flowId}`
    const resposta = await fetch(url);
    const dados = await resposta.json() as FlowInterface;
    return dados
}

async function finishNode(node?: NodeInterface, output?: object, error?: string) {
    const flowId = process.env.FLOW_ID as string
    const nodeId = process.env.NODE_ID as string
    
    const urlFinish = baseURL + `/finish-node?flowId=${flowId}`
        const body = {
        flow_id: flowId,
        node_id: nodeId,
        next_node_id: node ? node?.outputNodes[0] as string : undefined,
        node_output: output ? JSON.stringify(output) : undefined,
        error_message: error ? error : undefined
    } as FinishInterface

    const resposta2 = await fetch(urlFinish, {
        method: 'PATCH',
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify(body)
    });
    const dados2 = await resposta2.json();

}




async function getStateFromFlow() {
    const flowId = process.env.FLOW_ID
    const nodeId = process.env.NODE_ID as string
    
    const FLOW_STATE = await getFlowState()
    const node = getNode(FLOW_STATE.nodes, nodeId)
    const nodeInput = node?.state?.input ? JSON.parse(node.state.input) as any : undefined 

    //get filename from flow state
    const fileName = nodeInput.fileName as string




    if(!node) 
        return await finishNode(node, undefined, "Node not found")
    
    await writeFile(fileName, "Hello Password")

    await finishNode(node, {password: "Hello Password"})

}

(async ()=>{
    await getStateFromFlow()
})()

File: internal/tests/nodes_go_to_test/nodeMid/nodeMidToTest.go
package main

import (
	"encoding/json"
	"log"
	"os"

	"github.com/luizrgf2/universal-flow/internal/tests/nodes_go_to_test"
)

func main() {
	flowState, err := nodes_go_to_test.GetFlowState()
	if err != nil {
		log.Fatal(err)
	}

	nodeID := os.Getenv("NODE_ID")
	node := nodes_go_to_test.GetNode(flowState.Nodes, nodeID)

	if node == nil {
		nodes_go_to_test.FinishNode("", nil, "Node not found")
		log.Fatal("Node not found")
	}

	var inputState map[string]interface{}
	if input, ok := node.State["input"]; ok && input != "" {
		err := json.Unmarshal([]byte(input), &inputState)
		if err != nil {
			nodes_go_to_test.FinishNode(node.OutputNodes[0], nil, "Error parsing input state")
			log.Fatal(err)
		}
	}

	fileName, ok := inputState["fileName"].(string)
	if !ok {
		nodes_go_to_test.FinishNode(node.OutputNodes[0], nil, "fileName not found in input state")
		log.Fatal("fileName not found in input state")
	}

	err = os.WriteFile(fileName, []byte("Hello Password"), 0644)
	if err != nil {
		nodes_go_to_test.FinishNode(node.OutputNodes[0], nil, "Error writing file")
		log.Fatal(err)
	}

	output := map[string]string{
		"password": "Hello Password",
	}

	var nextNodeID string
	if len(node.OutputNodes) > 0 {
		nextNodeID = node.OutputNodes[0]
	}

	err = nodes_go_to_test.FinishNode(nextNodeID, output, "")
	if err != nil {
		log.Fatal(err)
	}
}
